Updates to the Java and VS Code docs (#14207)
github/codeql: docs/codeql/codeql-for-visual-studio-code/using-the-codeql-model-editor.rst

:tocdepth: 1

.. _using-the-codeql-model-editor:

Using the CodeQL model editor
=============================

.. include:: ../reusables/beta-note-model-pack-editor-vsc.rst

You can view, write, and edit all types of CodeQL packs in Visual Studio Code using the CodeQL extension. The model editor is designed to help you model external dependencies of your codebase that are not supported by the standard CodeQL Libraries.

About the CodeQL model editor
-----------------------------

> The CodeQL model editor guides you through modeling the calls to external dependencies in your application or fully modeling all the public entry and exit points in an external dependency

Reviewer: ```suggestion
The CodeQL model editor guides you through modeling the calls to external dependencies in your application or fully modeling all the public entry and exit points in an external dependency.
```

When you open the model editor, it analyzes the currently selected CodeQL database and identifies where the application uses external APIs and all public methods. An external (or third party) API is any API that is not part of the CodeQL database you have selected.

The model editor has two different modes:

> - Application mode (default view): The editor lists each external framework used by the seelcted CodeQL database. When you expand a framework, a list of all calls to and from the external API is shown with the options available to model dataflow through each call. This mode is most useful for improving the CodeQL results for the specific codebase.

Reviewer: ```suggestion
- Application mode (default view): The editor lists each external framework used by the selected CodeQL database. When you expand a framework, a list of all calls to and from the external API is shown with the options available to model dataflow through each call. This mode is most useful for improving the CodeQL results for the specific codebase.
```

> - Application mode (default view): The editor lists each external framework used by the seelcted CodeQL database. When you expand a framework, a list of all calls to and from the external API is shown with the options available to model dataflow through each call. This mode is most useful for improving the CodeQL results for the specific codebase.

Author: 🤦🏻‍♀️ Thank you. You both picked up this typo.

----------------------------------

#. Open your CodeQL workspace in VS Code, for example, the vscode-codeql-starter workspace.
   If you haven't updated the ``ql`` submodule for a while, update it from ``main`` to ensure that you have the queries used to gather data for the model editor.
#. Open the CodeQL extension and select the CodeQL database that you want to model from the "Databases" section of the left side pane.
#. Use the command palette to run the “CodeQL: Open Model Editor (Beta)” command.
#. The CodeQL model editor will open in a new tab and run a series of telemetry queries to identify APIs in the code.
#. When the queries are complete, the APIs that have been identified are shown in the editor.

Modeling the calls your codebase makes to external APIs
-------------------------------------------------------

You typically use this method when you are looking at a specific codebase where you want to improve the precision of CodeQL results. This is usually when the codebase uses frameworks or libraries that are not supported by CodeQL but they are not used by other teams in your organization. 

#. Select the CodeQL database that you want to improve CodeQL coverage for.

> #. Display the CodeQL model editor, by default the editor runs in application mode, so the list of external APIs used by the selected codebase is shown.

Reviewer: ```suggestion
#. Display the CodeQL model editor. By default the editor runs in application mode, so the list of external APIs used by the selected codebase is shown.
```

   .. image:: ../images/codeql-for-visual-studio-code/model-application-mode.png
      :width: 800
      :alt: Screenshot of the "Application mode" view of the CodeQL model pack editor in Visual Studio Code showing three of the external frameworks used by the "sofa-jraft" codebase.

#. Click to expand an external API and view the list of calls from the codebase to the external dependency.
#. Click **View** associated with an API call or method to show where it is used in your codebase.

   .. image:: ../images/codeql-for-visual-studio-code/model-application-mode-expanded.png
      :width: 800
      :alt: Screenshot of the "Application mode" view of the CodeQL model pack editor in Visual Studio Code showing the calls to the "rocksdbjni" framework ready for modeling. The "View" option for the first call is highlighted with a dark orange outline.

#. When you have determined how to model the call or method, define the **Model type**.
#. The remaining fields are updated with available options:

   - **Source**: choose the **Output** element to model.
   - **Sink**: choose the **Input** element to model.
   - **Flow summary**: choose the **Input** and **Output** elements to model.

> #. Define the **Kind** of data flow for the model.

Reviewer: ```suggestion
#. Define the **Kind** of dataflow for the model.
```

?

#. When you have finished modeling, click **Save all** or  **Save** (shown at the bottom right of each expanded list of calls). The percentage of calls modeled in the editor is updated.

The models are stored in your workspace at ``.github/codeql/extensions/<codeql-model-pack>``, where ``<codeql-model-packe>`` is the name of the CodeQL database that you selected. That is, the name of the repository, hyphen, the language analyzed by CodeQL.

The models are stored in a series of YAML data extension files, one for each external API. For example:

- ``.github/codeql/extensions/sofa-jraft-java`` - model pack
    - ``models``
        - ``jmh-core.model.yml`` - model of calls to the ``jmh-core@1.20`` dependency.

>         - ``rocksdbjni.model.yml`` - model of calls to the ``rocksdbjni@7.7.3`` dependency.

Reviewer: Would the comments be better after `#`s?

Modeling the public API of a codebase
-------------------------------------

You typically use this method when you want to model a framework or library that your organization uses in more than one codebase. Once you have finished creating and testing the model, you can publish the CodeQL model pack to the GitHub Container Registry for your whole organization to use.

#. Select the CodeQL database that you want to model.

> #. Display the CodeQL model editor, by default the editor runs in application mode. Click **Model as dependency** to display dependency mode. The screen changes to show the public API of the framework or library.

Reviewer: ```suggestion
#. Display the CodeQL model editor. By default the editor runs in application mode. Click **Model as dependency** to display dependency mode. The screen changes to show the public API of the framework or library.
```

   .. image:: ../images/codeql-for-visual-studio-code/model-dependency-mode.png
      :width: 800
      :alt: Screenshot of the "Dependency mode" view of the CodeQL model pack editor in Visual Studio Code showing three of the packages published by the "sofa-jraft" codebase.

#. Click to expand a package and view the list of available methods.
#. Click **View** associated with a method to show its definition.

   .. image:: ../images/codeql-for-visual-studio-code/model-dependency-mode-expanded.png
      :width: 800
      :alt: Screenshot of the "Dependency mode" view of the CodeQL model pack editor in Visual Studio Code showing the public methods available in the "com.alipay.soft.jraft.option" package ready for modeling. The "View" option for the first method is highlighted with a dark orange outline.

#. When you have determined how to model the method, define the **Model type**.
#. The remaining fields are updated with available options:

   - **Source**: choose the **Output** element to model.
   - **Sink**: choose the **Input** element to model.
   - **Flow summary**: choose the **Input** and **Output** elements to model.

> #. Define the **Kind** of data flow for the model.

Reviewer: ```suggestion
#. Define the **Kind** of dataflow for the model.
```

?

#. When you have finished modeling, click **Save all** or  **Save** (shown at the bottom right of each expanded list of calls). The percentage of calls modeled in the editor is updated.

> The models are stored in your workspace at ``.github/codeql/extensions/<codeql-model-pack>``, where ``<codeql-model-packe>`` is the name of the CodeQL database that you selected. That is, the name of the repository, hyphen, the language analyzed by CodeQL.

Reviewer 2: ```suggestion
The models are stored in your workspace at ``.github/codeql/extensions/<codeql-model-pack>``, where ``<codeql-model-pack>`` is the name of the CodeQL database that you selected. That is, the name of the repository, hyphen, the language analyzed by CodeQL.
```

The models are stored in a series of YAML data extension files, one for each public method. For example:

- ``.github/codeql/extensions/sofa-jraft-java`` - model pack
    - ``models``
        - ``com.alipay.sofa.jraft.option.model.yml`` - model of public methods in this Java package.

>         - ``com.alipay.sofa.jraft.rhea.options.model.yml`` - model of public methods in this Java package.

Reviewer: As above suggestion re `#`s

Displaying the CodeQL model editor
----------------------------------

#. Open your CodeQL workspace in VS Code, for example, the vscode-codeql-starter workspace.
   If you haven't updated the ``ql`` submodule for a while, update it from ``main`` to ensure that you have the queries used to gather data for the model editor.
#. Open the CodeQL extension and select the CodeQL database that you want to model from the "Databases" section of the left side pane.
#. Use the command palette to run the “CodeQL: Open Model Editor (Beta)” command.
#. The CodeQL model editor will open in a new tab and run a series of telemetry queries to identify APIs in the code.
#. When the queries are complete, the APIs that have been identified are shown in the editor.

Modeling the calls your codebase makes to external APIs
-------------------------------------------------------

> You typically use this method when you are looking at a specific codebase where you want to improve the precision of CodeQL results. This is usually when the codebase uses frameworks or libraries that are not supported by CodeQL but they are not used by other teams in your organization. 

Reviewer 3: `but they are not used by other teams in your organization.` -> `and if the sourcecode of the framework or library is not included in the analysis`.
One could imagine a non-monolith architecture where a project has dependencies to internal packages, but where the source code of these are not included in the analysis.

TODO - EDIT THIS CONTENT!

Explain how to find the data extension files that you've created and test them. Also how to save to the right location in a GitHub repository for default and advanced setup to use.

About the CodeQL model editor
-----------------------------

The CodeQL model editor is a new feature in CodeQL for VS Code to support GitHub staff creating CodeQL models for libraries and frameworks written in Java and C#.

The editor takes a CodeQL database and runs some telemetry queries to identify uses of APIs that can be used to reason about the dataflow through the codebase. There are two modes of operation:

- Application mode: the editor identifies the external APIs used by the codebase. An external (or third party) API is any API that is not part of the CodeQL database you are analyzing. This mode is most useful for improving CodeQL results for the specific codebase.
- Framework mode: the editor identifies the publicly accessible APIs in the codebase. This mode is most useful for improving the CodeQL results for any codebases that use those APIs.

Setting up the CodeQL model editor
----------------------------------

> To set up the CodeQL model editor, you need to be using CodeQL for VS Code 1.8.7 or later with the following settings: 
> 
> .. code-block:: json
> 
>     "codeQL.canary": true, 
>     "codeQL.model.editor": true,
> 
> CHECK THIS ^^^
> 

Reviewer 4: These settings won't be needed!

```suggestion
```

Open the user settings editor (JSON) using the command palette (Ctrl/Cmd+Shift+P) and using the command “Preferences: Open User Settings (JSON)”, add these two settings to the file.

If you want to test the CodeQL model packs you generate in VS Code then this setting is also required:
   ``"codeQL.runningQueries.useExtensionPacks": true`` CHECK THIS

> WHAT DOES ``"codeQL.model.llmGeneration": true,`` do?

Reviewer 4: This is an unrelated setting which isn't needed too.

> WHAT DOES ``"codeQL.model.llmGeneration": true,`` do?

Reviewer 4: ```suggestion
```
